File: maths/greatest_common_divisor.go
package main

import (
	"fmt"
	"math"
)

func GreatestCommonDivisorRecursive(a, b int) int {
	if a == 0 {
		return int(math.Abs(float64(b)))
	}
	return GreatestCommonDivisorRecursive(b%a, a)
}

func GreatestCommonDivisorIterative(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return int(math.Abs(float64(a)))
}

func main() {
	a, b := 24, 40
	fmt.Printf("GCD Recursive (%d, %d) = %d\n", a, b, GreatestCommonDivisorRecursive(a, b))
	fmt.Printf("GCD Iterative (%d, %d) = %d\n", a, b, GreatestCommonDivisorIterative(a, b))
}

File: maths/greatest_common_divisor.cpp
#include <cmath>
#include <iostream>

int greatestCommonDivisorRecursive(int a, int b) {
  if (a == 0)
    return std::abs(b);
  return greatestCommonDivisorRecursive(b % a, a);
}

int greatestCommonDivisorIterative(int a, int b) {
  while (b != 0) {
    int temp = b;
    b = a % b;
    a = temp;
  }
  return std::abs(a);
}

int main() {
  int a = 24, b = 40;
  std::cout << "GCD Recursive (" << a << ", " << b
            << ") = " << greatestCommonDivisorRecursive(a, b) << std::endl;
  std::cout << "GCD Iterative (" << a << ", " << b
            << ") = " << greatestCommonDivisorIterative(a, b) << std::endl;
  return 0;
}
